modules/ROOT/pages/patterns/shortest-paths.adoc

```diff
@@ -538,3 +538,90 @@ RETURN [stop IN nodes(p) | stop.name] AS r1,
 
 2+d|Rows: 1
 |===
+
+
+[[planning-StatefulShortestPath-operator]]
+== Planning the StatefulShortestPath-operator
+The `StatefulShortestPath(All)` uses a unidirectional breadth first search algorithm, while the `StatefulShortestPath(Into)` uses a bidirectional breadth first search algorithm.
+See xref:planning-and-tuning/operators/operators-detail.adoc#query-plan-stateful-shortest-path-into[the operator-detail page] for a description of the `StatefulShortestPath(All)` and `StatefulShortestPath(Into)` operators.
+
+The planner decides between the two variants using this heuristic:
+
+* When both boundary nodes have been matched in a `MATCH` clause prior to the `MATCH` clause containing the shortest paths query, `StatefulShortestPath(Into)` is planned.
+* Otherwise, if both boundary nodes match to at most one data node, then `StatefulShortestPath(Into)` is planned.
+* Otherwise, `StatefulShortestPath(All)` is planned.
+
+Sometimes the planner cannot make reliable estimations about how many data nodes a pattern node will match.
+Consider using a xref:constraints/index.adoc#unique-node-property[uniqueness constraint] where applicable to help the planner getting more reliable estimates.
+
+As an example, consider the case where the left boundary node matches to 10 data nodes and the right boundary node to 3 data nodes.
+Using `StatefulShortestPath(Into)` requires 30 bidirectional breath first search (BFS) invocations.
+Using `StatefulShortestPath(All)` requires only 3 unidirectional BFS invocations expanding from the right boundary node.
+
+This heuristic decision to favour `StatefulShortestPath(All)` can sometimes lead to worse query performance.
```

Comment:
```suggestion
However, the heuristic decision to favor `StatefulShortestPath(All)` can lead to worse query performance.
To have the planner choose the `StatefulShortestPath(Into)`, rewrite the query using a xref:subqueries/call-subquery.adoc[`CALL` subquery].
This is because the `CALL` subquery is executed for each incoming row.
Thus, in the below query, both `a` and `b` are matched to exactly one `Station` node respectively which enables the planner to use `StatefulShortestPath(Into)`.

.Query rewritten to use `StatefulShortestPath(Into)`
[source,cypher]
----
MATCH
  (a:Station),
  (b:Station)
CALL {
  WITH a, b
  MATCH
    p = ALL SHORTEST (a)(()-[]-()-[]-()){1,}(b)
  RETURN p
}
RETURN p
----

.Result
role="queryplan", subs="attributes+"]
----
+-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| Operator                    | Id | Details                                                                                 | Estimated Rows | Rows | DB Hits | Memory (Bytes) | Page Cache Hits/Misses | Time (ms) | Pipeline      |
+-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +ProduceResults             |  0 | p                                                                                       |            383 |  154 |       0 |              0 |                    0/0 |     0.166 |               |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +Projection                 |  1 | (a) ((anon_12)-[anon_14]-(anon_13)-[anon_11]-())* (b) AS p                              |            383 |  154 |       0 |                |                    0/0 |     2.605 |               |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+               |
| +StatefulShortestPath(Into) |  2 | SHORTEST 1 GROUPS (a) ((`anon_5`)-[`anon_6`]-(`anon_7`)-[`anon_8`]-(`anon_9`)){1, } (b) |            383 |  154 |    1517 |          24486 |                    1/0 |    76.973 | In Pipeline 3 |
| |                           |    |         expanding from: a                                                               |                |      |         |                |                        |           |               |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +CartesianProduct           |  3 |                                                                                         |            100 |   81 |       0 |           2056 |                    0/0 |     1.352 | In Pipeline 2 |
| |\                          +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| | +NodeByLabelScan          |  4 | b:Station                                                                               |             10 |    9 |      10 |            392 |                    1/0 |     0.229 | In Pipeline 1 |
| |                           +----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
| +NodeByLabelScan            |  5 | a:Station                                                                               |             10 |    9 |      10 |            376 |                    1/0 |     1.224 | In Pipeline 0 |
+-----------------------------+----+-----------------------------------------------------------------------------------------+----------------+------+---------+----------------+------------------------+-----------+---------------+
----
```